oauth-common/src/main/java/io/strimzi/kafka/oauth/common/PrincipalExtractor.java

```diff
@@ -88,4 +105,60 @@ public String toString() {
     public boolean isConfigured() {
         return usernameClaim != null || fallbackUsernameClaim != null || fallbackUsernamePrefix != null;
     }
+
+    private static List<String> parseClaimSpec(String spec) {
+        spec = spec == null ? null : spec.trim();
+        if (spec == null || "".equals(spec)) {
+            return null;
+        }
+
+        if (!spec.startsWith("[")) {
+            return Collections.singletonList(spec);
+        }
+
+        LinkedList<String> parsed = new LinkedList<>();
+        int pos = 0;
+        int epos = 0;
+        while (pos != -1) {
+            pos += 1;
+            epos = spec.indexOf("]", pos);
+            if (epos == -1) {
+                throw new IllegalArgumentException("Failed to parse username claim spec: '" + spec + "' (Missing ']')");
+            }
+            parsed.add(removeQuotesAndSpaces(spec.substring(pos, epos)));
```

Comment:
How does this handle the input `['foo'bar']`?

Reply:
I accept that this is an edge case, but I find it really surprising anything should accept `'foo'bar'` as if it were a literal containing the characters `foo'bar`. I can't think of any examples (e.g. in programming languages) where things are parsed where such input wouldn't be rejected. Such parsing is usually based on tokenisation and the idea of paired delimiters. If people need to include `'` in the middle of a literal like that the right way to do it is require escaping. You're violating a very very common expectation here. That would be OK if there was a valid use for supporting this, but you agree that no one should want or need this flexibility. 

I think it's far safer to reject inputs like this, since it's more than likely due to a user messing up their config than something intentional. And rejection would appear to be easy enough: Anything inside the quotes should match `[a-zA-Z0-9.]+` or similar.

In general it's much easier to be strict from that start, and relax once its clear that the rules are too strict for valid use cases that users have, than to try to restrict inputs later on. By that point you don't know what weird and wonderful things people might be using and expecting to continue working.

Reply:
@tombentley I modified the implementation to use JsonPath for nested attributes. This should remove any inconsistencies in special character matching ...